test/functorch/test_aotdispatch.py

```diff
@@ -2735,6 +2735,148 @@ def forward(self, *args):
+
+        mod = TestMod(fn)
+        inp = torch.randn(2, 2)
+
+        gm, _ = aot_export_module(mod, [inp], trace_joint=False, pre_dispatch=True)
+        self.assertExpectedInline(str(gm.code).strip(), """\
+def forward(self, arg0_1, arg1_1):
+    _set_grad_enabled = torch._C._set_grad_enabled(True)
+    mm = torch.ops.aten.mm.default(arg1_1, arg1_1)
+    _set_grad_enabled_1 = torch._C._set_grad_enabled(False)
+    add = torch.ops.aten.add.Tensor(mm, 2);  mm = None
+    sum_1 = torch.ops.aten.sum.default(arg1_1);  arg1_1 = None
+    sum_2 = torch.ops.aten.sum.default(add);  add = None
+    add_1 = torch.ops.aten.add.Tensor(sum_1, sum_2);  sum_1 = sum_2 = None
+    return (add_1,)""")
+
+    def test_aot_export_predispatch_func_view(self):
+        def fn(p, x):
+            y = x @ x
+            y.add_(2)
+            return (x.sum() + y.view(1, 4).sum(),)
+
+        mod = TestMod(fn)
+        inp = torch.randn(2, 2)
+
+        gm, _ = aot_export_module(mod, [inp], trace_joint=False, pre_dispatch=True)
+        self.assertExpectedInline(str(gm.code).strip(), """\
+def forward(self, arg0_1, arg1_1):
+    mm = torch.ops.aten.mm.default(arg1_1, arg1_1)
+    add = torch.ops.aten.add.Tensor(mm, 2);  mm = None
+    sum_1 = torch.ops.aten.sum.default(arg1_1);  arg1_1 = None
+    view_1 = torch.ops.aten.view.default(add, [1, 4]);  add = None
+    sum_2 = torch.ops.aten.sum.default(view_1);  view_1 = None
+    add_1 = torch.ops.aten.add.Tensor(sum_1, sum_2);  sum_1 = sum_2 = None
+    return (add_1,)""")
+
+    def test_aot_export_predispatch_buffer_mutation_metadata(self):
+        class Foo(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.register_buffer('foo', torch.zeros(2, 2))
+
+            def forward(self, x):
+                self.foo.add_(4)
+                return (x.sum() + self.foo.sum(),)
+
+        inp = torch.randn(2, 2)
+
+        gm, graph_sig = aot_export_module(Foo(), [inp], trace_joint=False, pre_dispatch=True)
+        self.assertExpectedInline(str(gm.code).strip(), """\
+def forward(self, arg0_1, arg1_1):
+    add = torch.ops.aten.add.Tensor(arg0_1, 4);  arg0_1 = None
+    sum_1 = torch.ops.aten.sum.default(arg1_1);  arg1_1 = None
+    sum_2 = torch.ops.aten.sum.default(add)
+    add_1 = torch.ops.aten.add.Tensor(sum_1, sum_2);  sum_1 = sum_2 = None
+    return (add, add_1)""")
+        eager_mod = Foo()
+        output_1, output_2 = gm(torch.zeros(2, 2), inp)
+        eager_output = eager_mod(inp)
+        self.assertTrue(torch.allclose(output_2, eager_output[0]))
+
+        _, output_2 = gm(output_1, inp)
+        eager_output = eager_mod(inp)
+        self.assertTrue(torch.allclose(output_2, eager_output[0]))
+        self.assertTrue("foo" in graph_sig.buffers)
+        self.assertTrue(graph_sig.inputs_to_buffers["arg0_1"] == "foo")
+
+    def test_aot_export_predispatch_with_autograd_op(self):
+        def foo(p, x):
+            torch._C._set_grad_enabled(True)
```

Comment:
This is kinda risky since this test directly mutates global state (and could affect other tests).

I would wrap the whole thing in a try/finally:
```

try:
    old_state = torch._C._is_grad_enabled()
    run()
finally:
    torch._C._set_grad_enabled(old)